core/src/main/scala/org/apache/spark/crypto/CryptoCodec.scala

```diff
@@ -0,0 +1,144 @@
+   * a lossless operation (concatenation, addition, or XOR).
+   * @see http://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_.28CTR.29
+   * @param initIV
+   * @param counter
+   * @param IV
+   */
+  def calculateIV(initIV: Array[Byte], counter: Long, IV: Array[Byte])
+
+  /**
+   * @return Encryptor the encryptor
+   */
+  def createEncryptor: Encryptor
+
+  /**
+   * @return Decryptor the decryptor
+   */
+  def createDecryptor: Decryptor
+
+  /**
+   * Generate a number of secure, random bytes suitable for cryptographic use.
+   * This method needs to be thread-safe.
+   * @param bytes byte array to populate with random data
+   */
+  def generateSecureRandom(bytes: Array[Byte])
+}
+
+object CryptoCodec extends Logging {
+  def getInstance(conf: SparkConf): CryptoCodec = {
+    var name: String = conf.get(SPARK_SECURITY_CRYPTO_CIPHER_SUITE_KEY,
+      SPARK_SECURITY_CRYPTO_CIPHER_SUITE_DEFAULT)
+    getInstance(conf, CipherSuite.convert(name))
+  }
+
+  def getInstance(conf: SparkConf, cipherSuite: CipherSuite): CryptoCodec = {
+    var klasses: List[String] = getCodecClasses(conf, cipherSuite)
+    var codec: CryptoCodec = null
+    for (klass <- klasses) {
+      try {
+        val m = universe.runtimeMirror(getClass.getClassLoader)
+        var c: CryptoCodec = null
+        if (klass.equals("org.apache.spark.crypto.JceAesCtrCryptoCodec")) {
+          val classCryptoCodec = universe.typeOf[org.apache.spark.crypto.JceAesCtrCryptoCodec]
+            .typeSymbol.asClass
+          val cm = m.reflectClass(classCryptoCodec)
+          val ctor = universe.typeOf[org.apache.spark.crypto.JceAesCtrCryptoCodec].declaration(
+            universe.nme.CONSTRUCTOR).asMethod
+          val ctorm = cm.reflectConstructor(ctor)
+          val p = ctorm(conf)
+          c = p.asInstanceOf[org.apache.spark.crypto.CryptoCodec]
+        } else {
+          // TODO add Openssl crypto codec
+        }
+
+        if (c.getCipherSuite.name.equals(cipherSuite.name)) {
+          if (codec == null) {
+            logDebug(s"Using crypto codec $klass.getName.")
+            codec = c
+          }
+        }
+        else {
+          logDebug(s"Crypto codec $klass.getName doesn't meet the cipher suite $cipherSuite" +
+            s".getName.")
+        }
+      }
+      catch {
+        case e: Exception => {
+          logDebug(s"Crypto codec $klass.getName is not available.")
+        }
+      }
+    }
+    codec
+  }
+
+  def getCodecClasses(conf: SparkConf, cipherSuite: CipherSuite): List[String] = {
+    var result: List[String] = List()
+    val configName: String = SPARK_SECURITY_CRYPTO_CODEC_CLASSES_AES_CTR_NOPADDING_KEY
+    val codecString: String = conf.get(configName)
+    if (codecString == null) {
+      logDebug("No crypto codec classes with cipher suite configured.")
+      null
+    }
+
+    var codecArray: Array[String] = codecString.trim.split(",")
+    for (c <- codecArray) {
```

Comment:
How is this different than just returning `codecString.trim.split(",").toList`?

You also have a try/catch but the exceptions are never thrown anywhere.